# NPHL Mpox WGS Analysis

This repository contains scripts to assist the **Cameroon NPHL Bioinformatic Team** in analyzing Mpox WGS data for outbreak tracking and characterization.

## Project Structure

```bash
.
├── LICENSE
├── phylogeography => Scripts for phylogenetic analysis, molecular dating, and migration plotting.
│   ├── AncestralChanges.py
│   ├── baltic.py
│   ├── compute_float_date.sh
│   ├── environment.yml
│   ├── final_DataViz.py
│   ├── Plot_migrations.py
│   └── run_phylogenetic_tree.sh
├── README.md
└── viralrecon_MPOX => Wrapper for the nf-core/viralrecon pipeline (assembly & consensus).
    └── run_nf_core_viralrecon.sh
```

## Installation & Requirements

### 1. nf-core/Viralrecon Environment

Ensure you have [Nextflow](https://www.nextflow.io/docs/latest/install.html) installed

```bash
nextflow pull nf-core/viralrecon -r 2.6.0
```

### 2. Phylogeography Environment

Ensure you have [Conda](https://docs.conda.io/en/latest/) installed.

``` bash
conda env create -f phylogeography/environment.yml
```

## Usage

### 1. Launch nf-core/Viralrecon

Use this step to identify circulating strains and build consensus sequences from raw reads.

``` bash
#Launch the script 
bash viralrecon_MPOX/run_nf_core_viralrecon.sh <DATA_DIR> <OUT_DIR> <MODE> [REF_FASTA] [REF_GFF]
```

 **Parameters:**

* `DATA_DIR`: Path to folder containing FASTQ files (required).
* `OUT_DIR`: Path where results will be saved (required).
* `MODE`:
  * `0` = Use built-in reference (NC_063383.1).
  * `1` = Provide custom FASTA/GFF (requires 4th and 5th arguments).
  * Default=`0`.
* `REF_FASTA` : Path to the reference `.fasta` file.
* `REF_GFF` : Path to the reference `.gff` file.

**NOTE :**\
Check `viralrecon_MPOX/run_nf_core_viralrecon.sh` lines 39-40 to ensure `R1_EXT` and `R2_EXT` match your `.fastq` file file extensions. Default values are :

``` bash
R1_EXT='_R1_001.fastq.gz'
R2_EXT='_R2_001.fastq.gz'
```

### 2. Phylogeography Analysis

This module performs alignment, phylogeny (IQ-TREE), and molecular dating (TreeTime).

``` bash
#Ensure environment is active 
conda activate phylodynamic

#Run the analysis pipeline
bash phylogeography/run_phylogenetic_tree.sh \
      <Sequences> \
      <Reference Genome> \
      <Date File> \
      <Locations File> \
      <Last Sample Date> \
      <output_directory>
```

**Parameters:**

* **Sequences:** Path to the FASTA file with all samples sequences.
* **Reference Genome:** Path to the reference genome in FASTA format.
* **Date File:** Path to a CSV/TSV file with columns: `name`, `date` (YYYY-MM-DD).
* **Location File:** Path to a CSV/TSV file with columns: `name`, `country`.
* **Last Sample Date:** The date of the last sample in the tree with the format `YYYY-MM-DD`
* **Output Directory:** Path to a Output Directory.

### 3. Visualization (Migration Plots)

Visualize viral introductions and migration events based on the phylogeographic analysis.

``` bash
#Ensure environment is active
conda activate phylodynamic

#Run the visualization script
python phylogeography/final_DataViz.py \
      --migration <OUTDIR>/mugration/mugration_results.csv \
      --pointsGeoloc <path/to/gps_coordinates.csv>

#See script usage
python final_DataViz.py -h
```

**Arguments for `final_DataViz.py`:**

* `--migration`: Path to `mugration_results.csv` generated by the phylogeography step.
* `--pointsGeoloc`: Path to a CSV with columns: `location`, `long`, `lat`.
* `--origins` (Optional): Filter by origin location (e.g., `--origins South Center`).
* `--destinations` (Optional): Filter by destination (e.g., `--destinations North-West Littoral East`).
